run-build-functions.sh

```diff
@@ -109,7 +113,42 @@ restore_node_modules() {
     #   (...)
     # }
     # We need to cache all the node_module dirs, or we'll always be installing them on each run
-    mapfile -t package_locations <<< "$(echo "$workspace_output" | jq -r '.data | fromjson | to_entries | .[].value.location')"
+    mapfile -t package_locations <<<"$(echo "$workspace_output" | jq -r '.data | fromjson | to_entries | .[].value.location')"
+  else
+    # In yarn 1.x workspaces only work on private repos.
+    # See https://classic.yarnpkg.com/lang/en/docs/workspaces/#toc-how-to-use-it
+    # we have to check the error message to see if the repo is public and uses
+    # yarn 2 (or greater) with workspaces
+    if [ $workspace_exit_code -eq 1 ]
+    then
+      workspace_error_message=$(jq -r ".data" "$workspace_errors_logfile" 2>/dev/null)
+      package_manager_version=$(yarn --version)
+      if [ "${package_manager_version:0:1}" -gt 1 ] && [ "$workspace_error_message" = "Workspaces can only be enabled in private projects." ]
```

Comment:
Since we know the `installer` up ahead and we can check for yarn's version (similar to what we're doing in here already) I was thinking we could simplify things here a bit and not require us to check the error logs of the first command execution? We could instead check:
- if the installer is yarn and we're using version greater than 1, run the remainder of this `then` block
- else run the `YARN_IGNORE_PATH=1` block.

Something like (nvm the chaos and indentation):
```bash
local package_manager_version
package_manager_version=$(yarn --version)

if [ "$installer" = "yarn"] && [ "${package_manager_version:0:1}" -gt 1 ]
then
  # the yarn berry workspaces logic
  (...)
else
  (...)
  workspace_output="$(YARN_IGNORE_PATH=1 yarn workspaces --json info 2>/dev/null)"
  (etc.)
fi
# Handle the output of the executions above
if [ ${#package_locations[@]} -ne 0 ]
  then
    echo "$installer workspaces detected"
    # Extract all the packages and respective locations.
    restore_js_workspaces_cache "${package_locations[@]}"
  else
    echo "No $installer workspaces detected"

Reply:
@JGAntunes Hey 👋

sorry for replying 4 days late, I forgot about this issue... :sweat_smile:  Your suggestion seems to be reasonable, I'd be happy to try it and add  add a test this week, if that's okay for you (tbh I might have to work on it on the weekend, because currently it doesn't fit into our schedule at work).

Reply:
@fulopkovacs sure thing, however works better for you 👍 again, apologies for taking so long to follow up.